aten/src/ATen/native/cpu/SortingKernel.cpp

```diff
@@ -108,32 +113,75 @@ static void sort_kernel(
       int64_t dim_size
     ) {
       using scalar_t = typename std::remove_pointer<decltype(values)>::type;
+      using accscalar_t = typename AccType<scalar_t>::type;
       auto values_accessor = StridedRandomAccessor<scalar_t>(
         values, values_dim_stride);
       auto indices_accessor = StridedRandomAccessor<int64_t>(
         indices, indices_dim_stride);
-      auto composite_accessor = CompositeRandomAccessorCPU<
-        decltype(values_accessor), decltype(indices_accessor)
-      >(values_accessor, indices_accessor);
-
-      if (descending) {
-        if (stable) {
-          std::stable_sort(composite_accessor, composite_accessor + dim_size,
-            KeyValueCompDesc<scalar_t>());
+      // If the dimension of the sort axis is not physically dense 
+      // or the data type of values is BFloat16, 
+      // allocate a new array of accscalar_t data type and copy data in.
+      if(values_dim_stride != 1 || std::is_same<scalar_t, BFloat16>::value) {
+        accscalar_t *accvalue = new accscalar_t[dim_size];
+        int64_t *accvalue_indices = new int64_t[dim_size];
+        for (int64_t j = 0; j < dim_size; j++) {
+          accvalue[j] = values_accessor[j];
+          accvalue_indices[j] = indices_accessor[j];
         }
```

Comment:
value and indices accessors are random access Iterators, so you can compress this code with `std::copy`.

Reply:
Also, the whole point of StridedAccessor and CompositeAccessor is to avoid any memory copies, which could be a lot if `numel() / dim_size` is pretty big. What about a different heuristic, that is we indeed make a copy but only if `dim_size > numel / dim_size`? @ngimel , what do you think?

Reply:
@jiayisunx , could you please also run some benchmarks on large batched tensors where you sort across all the dimensions? Some shapes to check could be `(1024, 1024, 1024), (1024 * 1024, 1024), (1024, 1024 * 1024)`.

Reply:
Sorry for the late reply, this part is optimized for the cases that the stride of the sorting dimension is not 1 or the data type of values is BFloat16, and in this case, we will allocate a new array of accscalar_t data type and copy the values into it, and make the sorting dimension memory continuous.

Reply:
I am not convinced this optimization with stride > 1 does not depend on how many elements you have to actually copy and or sort. I am not sure it is always an optimal behavior. Therefore it would be great to seem more comprehensive testing with different ratios of `shape[i] / prod(shape[:i] + shape[i+1:]` and different shape magnitudes. The benchmarks you posted are only for relatively small 2d tensors only.

Reply:
Simple example, imagine the the sorting dimension has size, say 2, while the rest of the tensor contains a lot more elements. Then, at each index advancement into the tensor we will need to make a copy, which is not obvious to be faster than sorting a non-contiguous tensor without any copies.

Reply:
Also, you can create a buffer tensor to copy into to eliminate the need for `new` all together. There is `at::get_thread_num` function to get the id of the working thread for proper indexing into the buffer.

Reply:
I have attached the benchmark data for the large batched tensors and sorted across all the dimensions, including these shapes (1024, 1024, 1024), (1024 * 1024, 1024), (1024, 1024 * 1024).

Reply:
For completeness let's also test small multi-dim tensors. If you get rid of `new`, I am going to be convinced :)

Reply:
I have attached the benchmark data for small multi-dim tensors and when the sorted dimension is small, including these shapes (10, 1024, 1024) dim=0, (1024, 10, 1024) dim=1, (1024, 1024, 10) dim=2, (10, 512, 512, 512) dim=0, (512, 10, 512, 512) dim=1, (512, 512, 10, 512) dim=2, (512, 512, 512, 10) dim=3.

Reply:
Thank you, great results! Let's remove new and use buffers instead, remove for-loop, and it is going to be ready for merge.

Reply:
As I noted here, allocating a new array and the for-loop has two purpose, for converting to accscalar_t data type and making the sorting dimension memory contiguous in the new array. Actually I didn't catch your meaning, how do you want me to reach these two purpose?

Reply:
You do need to reallocated buffers you copy into every single time indices advance into other dimensions but the one to be sorted. It is sufficient to have a single contiguous buffer per thread to copy into, sort it, and then copy back. A buffer is a single tensor of dimension num_threads times dim_size that could be created before and passed into the _dim_apply method that sort is using. Does it make more sense now?